src/rpc/src/chainStorageWatcher.ts

```diff
@@ -0,0 +1,167 @@
+  refreshLowerBoundMs: number,
+  refreshUpperBoundMs: number,
+) =>
+  Math.floor(Math.random() * (refreshUpperBoundMs - refreshLowerBoundMs)) +
+  refreshLowerBoundMs;
+
+const makePathSubscriber = <T>(onUpdate: UpdateHandler<T>) => ({
+  onUpdate,
+});
+
+/**
+ * Periodically queries the most recent data from chain storage, batching RPC
+ * requests for efficiency.
+ *
+ * @param rpcAddr RPC server URL
+ * @param unserialize CapData unserializer to use
+ * @returns
+ */
+export const makeAgoricChainStorageWatcher = (
+  rpcAddr: string,
+  unserialize: Unserialize<string>,
+  newPathQueryDelayMs = defaultNewPathQueryDelayMs,
+  refreshLowerBoundMs = defaultRefreshLowerBoundMs,
+  refreshUpperBoundMs = defaultRefreshUpperBoundMs,
+) => {
+  // Map of paths to [identifier, value] pairs of most recent response values.
+  //
+  // The 'identifier' is used to avoid notifying subscribers for already-seen
+  // values. For 'data' queries, 'identifier' is the blockheight of the
+  // response. For 'children' queries, 'identifier' is the stringified array
+  // of children.
+  const latestValueCache = new Map<string, [string, unknown]>();
+
+  const watchedPathsToSubscribers = new Map<string, Set<Subscriber<unknown>>>();
+  let isNewPathWatched = false;
+  let isQueryInProgress = false;
+  let nextQueryTimeout: number | null = null;
+
+  const queueNextQuery = () => {
+    if (isQueryInProgress || !watchedPathsToSubscribers.size) {
+      return;
+    }
+
+    if (isNewPathWatched) {
+      // If there's new paths to watch, schedule another query very soon.
+      if (nextQueryTimeout) {
+        window.clearTimeout(nextQueryTimeout);
+      }
+      nextQueryTimeout = window.setTimeout(queryUpdates, newPathQueryDelayMs);
+    } else {
+      // Otherwise, refresh after a normal interval.
+      nextQueryTimeout = window.setTimeout(
+        queryUpdates,
+        randomRefreshPeriod(refreshLowerBoundMs, refreshUpperBoundMs),
+      );
+    }
+  };
+
+  const queryUpdates = async () => {
+    isQueryInProgress = true;
+    nextQueryTimeout = null;
+    isNewPathWatched = false;
+
+    const paths = [...watchedPathsToSubscribers.keys()].map(pathKey =>
+      JSON.parse(pathKey),
+    );
+
+    if (!paths.length) {
+      isQueryInProgress = false;
+      return;
+    }
+
+    try {
+      const data = await batchVstorageQuery(rpcAddr, unserialize, paths);
+      watchedPathsToSubscribers.forEach((subscribers, path) => {
+        const { blockHeight, value } = data[path];
+        const lastValue = latestValueCache.get(path);
+
+        if (
+          lastValue &&
+          (blockHeight === lastValue[0] ||
+            JSON.stringify(value) === lastValue[0])
+        ) {
+          // The value isn't new, don't emit.
+          return;
+        }
+
+        latestValueCache.set(path, [
+          blockHeight ?? JSON.stringify(value),
+          value,
+        ]);
+
+        subscribers.forEach(s => {
+          s.onUpdate(harden(value));
+        });
```

Comment:
prefer control structures.
```suggestion
        for (const s of subscribers) {
          s.onUpdate(harden(value));
        }
```